src/macos_native/ffi.rs

```diff
@@ -0,0 +1,469 @@
+    dictionary::{CFDictionary, CFDictionaryRef, CFMutableDictionaryRef},
+    mach_port::CFIndex,
+    number::CFNumber,
+    runloop::{CFRunLoop, CFRunLoopRef},
+    set::{CFSet, CFSetGetValues, CFSetRef},
+    string::{CFString, CFStringRef},
+    ConcreteCFType,
+};
+use libc::c_void;
+use mach2::{
+    kern_return::{kern_return_t, KERN_SUCCESS},
+    port::MACH_PORT_NULL,
+};
+
+pub use self::io_hid_device::{IOHIDDevice, IOHIDDeviceRef};
+pub use self::io_hid_manager::{IOHIDManager, IOHIDManagerRef};
+
+// Keys are described in IOHIDDeviceKeys.h
+
+pub const kIOHIDVendorIDKey: &str = "VendorID";
+pub const kIOHIDProductIDKey: &str = "ProductID";
+pub const kIOHIDSerialNumberKey: &str = "SerialNumber";
+pub const kIOHIDManufacturerKey: &str = "Manufacturer";
+pub const kIOHIDProductKey: &str = "Product";
+pub const kIOHIDVersionNumberKey: &str = "VersionNumber";
+pub const kIOHIDTransportKey: &str = "Transport";
+pub const kIOHIDDeviceUsagePairsKey: &str = "DeviceUsagePairs";
+
+/// Default allocator for CoreFoundation.
+///
+/// See <https://developer.apple.com/documentation/corefoundation/kcfallocatordefault?language=objc>
+pub const kCFAllocatorDefault: CFAllocatorRef = std::ptr::null_mut();
+
+/// Default mach port for communication with IOKit.
+///
+/// See <https://developer.apple.com/documentation/iokit/kiomainportdefault?language=objc>
+pub(crate) const kIOMainPortDefault: mach_port_t = 0;
+
+/// Seperate module for the IOHIDManager type,
+/// so that we can allow non-snake case names.
+#[allow(non_snake_case)]
+mod io_hid_manager {
+    use core_foundation::{
+        base::TCFType, declare_TCFType, impl_CFTypeDescription, impl_TCFType, mach_port::CFTypeID,
+    };
+    use std::os::raw::c_void;
+
+    #[repr(C)]
+    pub struct __IOHIDManager(c_void);
+    pub type IOHIDManagerRef = *const __IOHIDManager;
+
+    declare_TCFType!(IOHIDManager, IOHIDManagerRef);
+    impl_TCFType!(IOHIDManager, IOHIDManagerRef, IOHIDManagerGetTypeID);
+    impl_CFTypeDescription!(IOHIDManager);
+
+    extern "C" {
+        fn IOHIDManagerGetTypeID() -> CFTypeID;
+    }
+}
+
+impl IOHIDManager {
+    pub fn create() -> Self {
+        let manager = unsafe { IOHIDManagerCreate(kCFAllocatorDefault, 0) };
+
+        // TODO: Error handling
+        assert!(!manager.is_null());
+
+        unsafe { IOHIDManager::wrap_under_create_rule(manager) }
+    }
+
+    pub fn set_device_matching(&self, matching: Option<&CFDictionary<CFString, CFNumber>>) {
+        unsafe {
+            IOHIDManagerSetDeviceMatching(
+                self.as_concrete_TypeRef(),
+                matching
+                    .map(|m| m.as_concrete_TypeRef())
+                    .unwrap_or(std::ptr::null()),
+            );
+        }
+    }
+
+    pub fn copy_devices(&self) -> Vec<IOHIDDevice> {
+        let set: CFSet<IOHIDDeviceRef> = unsafe {
+            let set = IOHIDManagerCopyDevices(self.as_concrete_TypeRef());
```

Comment:
This isn't documented, but `IOHIDManagerCopyDevices` can return `nullptr` if there are no matching devices.